sktime/annotation/eagglo.py

```diff
@@ -0,0 +1,430 @@
+
+        # set up left and right neighbors
+        # special case for clusters 0 and n_cluster-1 to allow for cyclic merging
+        self.left = np.zeros(2 * self.n_cluster - 1, dtype=int)
+        self.left[: self.n_cluster] = [
+            i - 1 if i >= 1 else self.n_cluster - 1 for i in range(self.n_cluster)
+        ]
+
+        self.right = np.zeros(2 * self.n_cluster - 1, dtype=int)
+        self.right[: self.n_cluster] = [
+            i + 1 if i + 1 < self.n_cluster else 0 for i in range(self.n_cluster)
+        ]
+
+        # True means that a cluster has not been merged_
+        self.open = np.ones(2 * self.n_cluster - 1, dtype=bool)
+
+        # which clusters were merged_ at each step
+        self.merged_ = np.empty((self.n_cluster - 1, 2))
+
+        # set initial gof_ value
+        self.gof_ = np.array(
+            [
+                sum(
+                    self.distances[i, self.left[i]] + self.distances[i, self.right[i]]
+                    for i in range(self.n_cluster)
+                )
+            ]
+        )
+
+        # change point progression
+        self.progression = np.empty((self.n_cluster, self.n_cluster + 1))
+        self.progression[0, :] = [
+            sum(self.sizes[:i]) if i > 0 else 0 for i in range(self.n_cluster + 1)
+        ]  # N + 1 for cyclic mergers
+
+        # array to specify the starting point of a cluster
+        self.lm = np.zeros(2 * self.n_cluster - 1, dtype=int)
+        self.lm[: self.n_cluster] = range(self.n_cluster)
+
+    def _gof_update(self, i: int) -> float:
+        """Compute the updated goodness-of-fit statistic, left cluster given by i."""
+        fit = self.gof_[-1]
+        j = self.right[i]
+
+        # get new left and right clusters
+        rr = self.right[j]
+        ll = self.left[i]
+
+        # remove unneeded values in the gof_
+        fit -= 2 * (
+            self.distances[i, j] + self.distances[i, ll] + self.distances[j, rr]
+        )
+
+        # get cluster sizes
+        n1 = self.sizes[i]
+        n2 = self.sizes[j]
+
+        # add distance to new left cluster
+        n3 = self.sizes[ll]
+        k = (
+            (n1 + n3) * self.distances[i, ll]
+            + (n2 + n3) * self.distances[j, ll]
+            - n3 * self.distances[i, j]
+        ) / (n1 + n2 + n3)
+        fit += 2 * k
+
+        # add distance to new right
+        n3 = self.sizes[rr]
+        k = (
+            (n1 + n3) * self.distances[i, rr]
+            + (n2 + n3) * self.distances[j, rr]
+            - n3 * self.distances[i, j]
+        ) / (n1 + n2 + n3)
+        fit += 2 * k
+
+        return fit
+
+    def _find_closest(self, K: int) -> Tuple[int, int]:
+        """Determine which clusters will be merged_, for K clusters.
+
+        Greedily optimize the goodness-of-fit statistic by merging the pair of adjacent
+        clusters that results in the largest increase of the statistic's value.
+
+        Parameters
+        ----------
+        K: int
```

Comment:
Sorry if I am being pedantic here, but I do not think `K` is as you say it is. Since where this is called `K` is in the range (self.n_cluster - 1, 2 * self.n_cluster - 2).  It is related to the number of original clusters yes... My suggestion would be to actually have it be range(self.n_cluster) and do additions needed to make it easier for indexing here. Just found this part slightly confusing as is.

Reply:
I'm not sure, I'm guessing this follows a convention from the paper/original R implementation.

Reply:
Maybe @KatieBuc could comment?